pkg/container/ring.go

```diff
@@ -224,37 +260,38 @@ func (r *Ring) readFrom(ctx context.Context, read uint64) <-chan *v1.Event {
 			// write: f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
 			// index:  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
 			// cycle: 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f 1f  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
-			case readCycle >= (writeCycle+1)&r.cycleMask && readCycle < (halfCycle+writeCycle)&r.cycleMask:
+			case event == nil || readCycle >= (writeCycle+1)&r.cycleMask && readCycle < (halfCycle+writeCycle)&r.cycleMask:
 				// The writer has already written a new event so there's no
 				// need to stop the reader.
-				//
-				// FIXME?: It is still possible the writer will
-				//  write something after we check for `r.write` and before
-				//  we do r.cond.Wait, which means we will only receive
-				//  a broadcast() from the writer after a new event arrives,
-				//  making the reader block.
-				//  We can sort of avoiding it by checking if r.write got
-				//  incremented before we block while we wait for a broadcast
+
+				// Before going to sleep, we need to check that there has been
+				// no write in the meantime. This check can only be race free
+				// if the lock on notifyMu is held, otherwise a write can occur
+				// before we obtain the notifyCh instance.
+				r.notifyMu.Lock()
 				if lastWrite != atomic.LoadUint64(&r.write)-1 {
+					// A write has occurred - retry
+					r.notifyMu.Unlock()
 					read--
 					continue
 				}
-				r.cond.L.Lock()
-				select {
-				case <-ctx.Done():
-					r.cond.L.Unlock()
-					return
-				default:
-					r.cond.Wait()
-					r.cond.L.Unlock()
-					read--
+
+				// This channel will be closed by the writer if it makes a write
+				if r.notifyCh == nil {
+					r.notifyCh = make(chan struct{})
```

Comment:
Isn't possible to simply send an empty struct into `notifyCh` rather than re-creating the channel every time?

Reply:
Unfortunately not. We need to close the channel since close is broadcast to all consumers. A simple write would only notify a single consumer.

Reply:
Then wouldn't [`sync.Cond`](https://golang.org/pkg/sync/#Cond) be more appropriate as, in effect, you want to broadcast to all consumers?

Reply:
This is what we had previously. Unfortunately it is not flexible enough to react to context cancellation (this seems to be a common issue, see https://github.com/golang/go/issues/16620):

From the commit message:
```
The usage of `sync.Cond` in `Ring.readFrom` was prone to races, i.e.
it could miss signals that should stop it from going to sleep.
The following two cases addressed by this commit:

1. The `readFrom` goroutine could to sleep right after `r.write` has
   been bumped, therefore missing the signal from `cond.Broadcast()`.
   The previous code has tried to mitigate this (see removed FIXME
   comment), however that check still suffered from TOCTOU and
   therefore was not sufficient. This is now addressed by using a mutex
   in the writer that is held between the time where `r.write` is
   updated and the signal is being sent. This mutex is not held by any
   readers which are not trying to read the current write pointer,
   therefore concurrency of readers with the writer is preserved for
   most cases.

2. The `readFrom` goroutine could also enter sleep without noticing
   that its context is cancelled right before `cond.Wait()` is called.
   While there was a goroutine which sent a signal when the context is
   cancelled, there was still a possibility for a race where this
   signal was sent before `cond.Wait()` was actually called. This is
   mitigated by replacing `sync.Cond` with a channel based notification
   mechanism that allows us to use a `select` statement to wait for both
   the writer notification and the context cancellation.
```

Reply:
OK let's do it this way.